File: fixtures/java/src/com/acme/order/Basket.java
package com.acme.order;

import java.util.ArrayList;
import java.util.List;

/** A basket being filled. Declares no constructor of its own. */
public class Basket {

    private long total;

    private final List<Line> lines = new ArrayList<>();

    public void add(Line line) {
        lines.add(line);
        total += line.amount();
    }

    public long total() {
        return total;
    }

    public Cursor cursor() {
        return new Cursor();
    }

    /** Static nested: no enclosing Basket instance. */
    public static class Line {

        private final long amount;

        public Line(long amount) {
            this.amount = amount;
        }

        public long amount() {
            return amount;
        }

        /** Nested inside a nested type. */
        public static class Discount {

            public long applyTo(long amount) {
                return amount / 2;
            }
        }
    }

    /** Non-static inner: reads the enclosing Basket's fields. */
    public class Cursor {

        private int index;

        public Line next() {
            return lines.get(index++);
        }
    }
}

File: fixtures/csharp/src/Acme/Order/Basket.cs
using System.Linq;

namespace Acme.Order;

/// <summary>A basket being filled. Declares no constructor of its own.</summary>
public class Basket
{
    private long total;

    private readonly List<Line> lines = new();

    public void Add(Line line)
    {
        lines.Add(line);
        total += line.Amount;
    }

    public long Total => total;

    /// <summary>An indexer: a property whose symbol is `Item(System.Int32)`.</summary>
    public Line this[int index] => lines[index];

    public Cursor GetCursor()
    {
        return new Cursor(this);
    }

    public int Count => lines.Count(line => line.Amount > 0);

    /// <summary>Nested type; C# nesting is always static.</summary>
    public class Line
    {
        public Line(long amount)
        {
            Amount = amount;
        }

        public long Amount { get; }

        /// <summary>Nested inside a nested type.</summary>
        public class Discount
        {
            public long ApplyTo(long amount)
            {
                return amount / 2;
            }
        }
    }

    /// <summary>Reads the enclosing basket's lines through an explicit reference.</summary>
    public class Cursor
    {
        private readonly Basket basket;

        private int index;

        public Cursor(Basket basket)
        {
            this.basket = basket;
        }

        public Line Next()
        {
            return basket.lines[index++];
        }
    }
}
